### PR TITLE
sql: add locality to system.sql_instances table

### PR DESCRIPTION
**Pull request by rharding6373**
This PR adds the column `locality` to the `system.sql_instances` table
that contains the locality (e.g., region) of a SQL instance. The encoded
locality is a string representing the `roachpb.Locality` that may have
been provided when the instance was created.

This change also pipes the locality through `InstanceInfo`. This will
allow us to determine and use locality information of other SQL
instances, e.g. in DistSQL for multi-tenant locality-awareness
distribution planning.

Informs: #80678

Release note (sql change): Table `system.sql_instances` has a new
column, `locality`, that stores the locality of a SQL instance if it was
provided when the instance was started. This exposes a SQL instance's
locality to other instances in the cluster for query planning.

### CODE REVIEW

**Review by rharding6373 — COMMENTED** [review]
First commit is #82844

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N4eMqn11IlG8B2MDt5A:b-uhhntv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 24 of 25 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rharding6373)

___
*[`pkg/sql/logictest/logic.go` line 2250 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gwfy2CQ2rDHpJ1fgG:-N4gwfy2CQ2rDHpJ1fgH:bt8wg4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88f52460ff78a00685d1f6b09c35bb50f2c9784/pkg/sql/logictest/logic.go#L2250)):*
> ```Go
> 		// Enumerate all the blocked configs if the blocked config is a default
> 		// config list.
> 		<<<<<<< HEAD
> ```

Looks like a merge conflict resolution in the wrong commit.
___
*[`pkg/sql/sqlinstance/instancestorage/instancestorage.go` line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gx6cI8kYgj8gzV4P_:-N4gx6cI8kYgj8gzV4Pa:b-eq54v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/sql/sqlinstance/instancestorage/instancestorage.go#L72)):*
> ```Go
> // CreateInstance allocates a unique instance identifier for the SQL pod and
> // associates it with its SQL address and session information.
> // TODO(harding): add locality info here? This seems to be where the system.sql_instances
> ```

nit: is the TODO still relevant?
___
*[`pkg/upgrade/upgrades/alter_sql_instances_locality.go` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gxSVV1PkUtmlLY6ZO:-N4gxSVV1PkUtmlLY6ZP:btizpu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/upgrade/upgrades/alter_sql_instances_locality.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

nit: `s/2021/2022/`.
___
*[`pkg/sql/sqlinstance/instancestorage/instancereader_test.go` line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gxwuv2ci5Qvon2EnP:-N4gxwuv2ci5Qvon2EnQ:b-z93zzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/sql/sqlinstance/instancestorage/instancereader_test.go#L100)):*
> ```Go
> 				}
> 				if !locality.Equals(instanceInfo.Locality) {
> 					return errors.Newf("expected instance locality %s != actual instance locality %s", locality.String(), instanceInfo.Locality.String())
> ```

nit: I think there is no need to call `String()` when using `%s` format directive.
___
*[`pkg/upgrade/upgrades/alter_sql_instances_locality_test.go` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gyHTMAiOnsm5Bf_7p:-N4gyHTMAiOnsm5Bf_7q:b-iipjm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/upgrade/upgrades/alter_sql_instances_locality_test.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

nit: ditto
___
*[`pkg/upgrade/upgrades/alter_sql_instances_locality_test.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gyKk40XFy8knIDeqd:-N4gyKk40XFy8knIDeqe:b-k10keo) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/upgrade/upgrades/alter_sql_instances_locality_test.go#L68)):*
> ```Go
> 	// Inject the old copy of the descriptor.
> 	upgrades.InjectLegacyTable(ctx, t, s, systemschema.SQLInstancesTable, getDeprecatedSqlInstancesDescriptor)
> 	// Validate that the table statistics table has the old schema.
> ```

nit: comment needs an update.

**Review by rharding6373 — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N4htQua0vhoddp88Km2:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/sql/logictest/logic.go` line 2250 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gwfy2CQ2rDHpJ1fgG:-N4hgOJaEYWPCnsljyP8:b-15if7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88f52460ff78a00685d1f6b09c35bb50f2c9784/pkg/sql/logictest/logic.go#L2250)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Looks like a merge conflict resolution in the wrong commit.
</blockquote></details>

Ah, thought I fixed this botched merge, but I guess I only fixed it in the last commit. Looks like it's resolved now.
___
*[`pkg/sql/sqlinstance/instancestorage/instancestorage.go` line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N4gx6cI8kYgj8gzV4P_:-N4hgyHB5QzqHzI6KGyc:b7hrw1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/405dab2bf91bcbb456a476b3219da87a94f0ea2e/pkg/sql/sqlinstance/instancestorage/instancestorage.go#L72)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: is the TODO still relevant?
</blockquote></details>

No, thanks for the catch.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 26 of 29 files at r1, 1 of 25 files at r4, 24 of 24 files at r5, 24 of 24 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N55uXrT9LQVRKtF8Fi6:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach)

**Review by postamar — APPROVED** [review]
I realize I was added as a reviewer mainly to vet the upgrade logic (which LGTM!), though I was a bit surprised at the use of a `STRING` column type and a corresponding pseudo-CSV string encoding for `roachpb.Locality`, instead of `JSONB`, or perhaps `BYTES` containing the proto binary encoding. I'm guessing that that choice is deliberate & it's not bad per se in my view; I just want to raise awareness that this encoding choice effectively prevents adding further levels of nesting to this `roachpb.Locality` type. It also forces the use of regexp functions should we want to extract a specific key in a SQL query. 

I have no idea whether this is a valid concern or not since I don't know how this type is effectively used. Feel free to disregard.

**Review by rharding6373 — COMMENTED** [review]
That's a good point, I didn't consider the long-term implications of using a string type when I was pushing this proto around. However, json doesn't preserve order which is a required property for roachpb.Locality. We seem to use our own json library, so maybe we can make an ordered version, but I'm not sure that would work well, either.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N5C8Hh61rKJW1C-4qc2:btw8yge)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @yuzefovich)

**Comment by knz** [comment]
What about a SQL string array `[]STRING`?

File: pkg/clusterversion/cockroach_versions.go [hunk 1]
```diff
@@ -381,6 +381,10 @@ const (
 	// Previously, SSTs containing these could error.
 	AddSSTableTombstones
```
Comment (cucaroach) [inline pkg/clusterversion/cockroach_versions.go:383]: nit: empty line not conventional maybe?

**Review by rharding6373 — COMMENTED** [review]
After thinking about this a bit more, I made the new column type a jsonb, but it currently only has a single field in the json for the single field in locality. This seems to me to be a reasonable compromise between future-proofing and maintaining ordering. Would appreciate someone taking a look. Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N5HCyWX4nzU52b1MsRQ:b21xs34)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rharding6373, and @yuzefovich)

**Review by cucaroach — APPROVED** [review]
> Reviewed 10 of 24 files at r6, 9 of 38 files at r7, 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rharding6373 and @yuzefovich)

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 1 of 38 files at r7, 4 of 6 files at r8, 12 of 12 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rharding6373 and @yuzefovich)

___
*[`pkg/sql/sqlinstance/instancestorage/row_codec.go` line 134 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N5bJh3A5Am74WeFT6OW:-N5bJh3A5Am74WeFT6OX:b364mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/383313feb9d0b73cc25b65d311183a1d44f1e3b0/pkg/sql/sqlinstance/instancestorage/row_codec.go#L134)):*
> ```Go
> 		sessionID = sqlliveness.SessionID(tree.MustBeDBytes(sessionIDVal))
> 	}
> 	locality = roachpb.Locality{}
> ```

nit: this line seems unnecessary.
___
*[`pkg/sql/sqlinstance/instancestorage/row_codec.go` line 139 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N5bJokGC_9gkZkbglX0:-N5bJokGC_9gkZkbglX1:b-811d14) ([raw file](https://github.com/cockroachdb/cockroach/blob/383313feb9d0b73cc25b65d311183a1d44f1e3b0/pkg/sql/sqlinstance/instancestorage/row_codec.go#L139)):*
> ```Go
> 		v, err := localityJ.FetchValKey("Tiers")
> 		if err != nil {
> 			return instanceID, "", "", roachpb.Locality{}, hlc.Timestamp{}, false, err
> ```

nit: maybe add some context to this error (like "failed to find Tiers attribute in Locality JSON") ?
___
*[`pkg/util/json/json.go` line 394 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N5bKFW9Cq33XDZqgvdL:-N5bKFW9Cq33XDZqgvdM:bj4sn5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/383313feb9d0b73cc25b65d311183a1d44f1e3b0/pkg/util/json/json.go#L394)):*
> ```Go
> // BuildUnsorted creates a JSON object that uniqueifies the keys, but leaves
> // the keys in their original order instead of sorting them by key.
> func (b *ObjectBuilder) BuildUnsorted() JSON {
> ```

Is this used anywhere? I can't seem to find references.

**Review by rharding6373 — COMMENTED** [review]
TFTR! I think I finally fixed all the tests, but I'm going to make sure the tests are green before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-:-N5kZ8vx3QXm5uGOdGiG:bkfcqy1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/util/json/json.go` line 394 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82915#-N5bKFW9Cq33XDZqgvdL:-N5kT50tBzepPc7-kj3b:b-qtzz1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/383313feb9d0b73cc25b65d311183a1d44f1e3b0/pkg/util/json/json.go#L394)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Is this used anywhere? I can't seem to find references.
</blockquote></details>

Thanks for catching this. It was part of an initial draft to try to get JSON to work for roachpb.locality, but I forgot to remove it.

**Pull request merged**